core/src/main/java/com/linecorp/armeria/server/ServerListenerBuilder.java

```diff
@@ -262,6 +266,64 @@ public ServerListenerBuilder whenStopped(Iterable<? extends Consumer<? super Ser
         return this;
     }
 
+    /**
+     * Add a callback that gracefully shuts down the given {@link ExecutorService} when the {@link Server}
+     * is stopping.
+     *
+     * <p>Depending on the value of {@code terminationTimeoutMillis}, the behavior is as follows:
+     * <ul>
+     * <li>If the {@code terminationTimeoutMillis} is positive, it allows a maximum duration of
+     * {@code terminationTimeoutMillis} for the {@link ExecutorService} to terminate.</li>
+     * <li>If the {@code terminationTimeoutMillis} is zero, it will wait indefinitely for the
+     * {@link ExecutorService} to terminate during shutdown.</li>
+     * <li>If the {@code terminationTimeoutMillis} is negative, throws {@link IllegalArgumentException}.</li>
+     * </ul>
+     */
+    @UnstableApi
+    public ServerListenerBuilder shutdownWhenStopping(ExecutorService executorService,
+                                                      Duration terminationTimeout) {
+        requireNonNull(executorService, "executorService");
+        requireNonNull(terminationTimeout, "terminationTimeout");
+        long millis;
+        try {
+            millis = terminationTimeout.toMillis();
+        } catch (ArithmeticException ignore) {
+            millis = terminationTimeout.isNegative() ? Long.MIN_VALUE : Long.MAX_VALUE;
+        }
+        shutdownWhenStopping(executorService, millis);
+        return this;
+    }
+
+    /**
+     * Add a callback that gracefully shuts down the given {@link ExecutorService} when the {@link Server}
+     * is stopping.
+     *
+     * <p>Depending on the value of {@code terminationTimeoutMillis}, the behavior is as follows:
+     * <ul>
+     * <li>If the {@code terminationTimeoutMillis} is positive, it allows a maximum duration of
+     * {@code terminationTimeoutMillis} for the {@link ExecutorService} to terminate.</li>
+     * <li>If the {@code terminationTimeoutMillis} is zero, it will wait indefinitely for the
+     * {@link ExecutorService} to terminate during shutdown.</li>
+     * <li>If the {@code terminationTimeoutMillis} is negative, throws {@link IllegalArgumentException}.</li>
+     * </ul>
+     */
+    @UnstableApi
+    public ServerListenerBuilder shutdownWhenStopping(ExecutorService executorService,
+                                                      long terminationTimeoutMillis) {
+        requireNonNull(executorService, "executorService");
+        if (terminationTimeoutMillis < 0) {
+            throw new IllegalArgumentException("terminationTimeout must be greater than or equal to 0.");
+        }
+
+        if (terminationTimeoutMillis == 0) {
```

Comment:
We can assume that `Long.MAX_VALUE` is an infinite number.
```suggestion
        if (terminationTimeoutMillis == 0 || terminationTimeoutMillis == Long.MAX_VALUE) {
```

Reply:
While `Long.MAX_VALUE` in milliseconds represents an extremely long time, would it be okay to distinguish between the two? The reason for wanting to distinguish them is that in the `shutdownWhenStopping(ExecutorService executorService, Duration terminationTimeout)` method, when converting `Duration` to `long`, there's a risk of overflow, and in such cases, `Long.MAX_VALUE` or `Long.MIN_VALUE` is being used."

```
    @UnstableApi
    public ServerListenerBuilder shutdownWhenStopping(ExecutorService executorService,
                                                      Duration terminationTimeout) {
        requireNonNull(executorService, "executorService");
        requireNonNull(terminationTimeout, "terminationTimeout");
        long millis;
        try {
            millis = terminationTimeout.toMillis();
        } catch (ArithmeticException ignore) {
            millis = terminationTimeout.isNegative() ? Long.MIN_VALUE : Long.MAX_VALUE;
        }
        shutdownWhenStopping(executorService, millis);
        return this;
    }
```

Reply:
How about also branching it inside of the `ShutdownSupport.of(ExecutorService executor, long terminationTimeoutMillis)` method so that we can have the logic in one place?
```java
serverStoppedCallbacks.add(s -> ShutdownSupport.of(executorService, terminationTimeoutMillis).shutdown());
...
serverStoppingCallbacks.add(s -> ShutdownSupport.of(executorService, terminationTimeoutMillis).shutdown());

static ShutdownSupport of(ExecutorService executor, long terminationTimeoutMillis) {
    requireNonNull(executor, "executor");
    if (terminationTimeoutMillis == 0 || terminationTimeoutMillis == Long.MIN_VALUE) {
        return of(executor);
    }
    ...
}
```